### PR TITLE
hasBasename bugfix

### PR DESCRIPTION
**Pull request by hmate9**
Previously it didn't work when `prefix` had special regex characters (eg `$`)

### CODE REVIEW

**Comment by mjackson** [comment]
I'm uncomfortable merging this because I think that using a regex in the underlying code is probably too complex for our needs. All we really need to do is make a string comparison that asserts the *next* char is one of the URL-delimiting characters (i.e. `/`, `?`, `#`) or the end of the string.

Would you be open to doing a more extensive refactoring here?

**Comment by Parakleta** [comment]
I use the following successfully:

```js
export const hasBasename = (path, prefix) =>
    path.startsWith(prefix) && "/?#".indexOf(path.charAt(prefix.length)) !== -1;
```

This obviously works in the delimited case, but in the end-of-string case it also works because the `charAt` function returns the empty string if the length is past the end, and `indexOf` always returns `0` when called with the empty string as an argument.

**Comment by hmate9** [comment]
@Parakleta I added your change to the PR, just also converting `path` and `prefix` to lowercase to make sure the matching is case insensitive.

**Comment by hmate9** [comment]
I don't know why GitHub says there is a merge conflict, there isn't. It is a one line change.

**Comment by hmate9** [comment]
@mjackson could you please comment on this PR. Is it ready to be merged?

File: modules/PathUtils.js [hunk 1]
```diff
@@ -5,7 +5,7 @@ export const stripLeadingSlash = path =>
   path.charAt(0) === "/" ? path.substr(1) : path
 
 export const hasBasename = (path, prefix) =>
-  new RegExp("^" + prefix + "(\\/|\\?|#|$)", "i").test(path)
+  path.toLowerCase().lastIndexOf(prefix.toLowerCase(), 0) === 0 && "/?#".indexOf(path.charAt(prefix.length)) !== -1
```
Comment (mjackson) [inline modules/PathUtils.js]: Is there a particular reason you're using `lastIndexOf` instead of `indexOf` here?

**Comment by mjackson** [comment]
Thanks for working on this, @hmate9. Looks great. Just curious about the usage of `lastIndexOf` instead of `indexOf`.

**Comment by Parakleta** [comment]
`lastIndexOf` with a `fromIndex` of `0` is functionally equivalent to `startsWith` in this case but is available everywhere (`lastIndexOf` is ES1, `startsWith` is ES6).  Alternatively you could use something like `path.substr(0, prefix.length) === prefix` which is ES3.

**Comment by hmate9** [comment]
@Parakleta Are you sure about that? According to the docs:

> The indexOf() method returns the index within the calling String object of the first occurrence of the specified value.

It wouldn't make sense to look for more than one match if you have already found one.

**Comment by Parakleta** [comment]
You’re assuming you find a match...

**Comment by hmate9** [comment]
@Parakleta cool, makes sense. Thank you. What do you think @mjackson ? Should I change it back to `lastIndexOf`?

**Comment by hmate9** [comment]
@mjackson could we get an update on this please?

**Comment by hmate9** [comment]
ping @mjackson 

This is ready to be merged and is pretty useful fix

**Comment by avrahams1** [comment]
Is this going to be merged soon?

**Comment by bjarkehs** [comment]
Any news on this? I just ran into this issue myself.

**Comment by ajhyndman** [comment]
I'm running into this too.  I also realized that there are 4 other PRs that have been opened since, attempting to fix the same bug.

#608 
#607 
#604 
#566

**Comment by marconi** [comment]
we just ran into this as well LGTM! 👍

**Comment by mjackson** [comment]
Thanks for the PR, @hmate9, and sorry for sitting on it for so long. I just merged this manually. It will be released in 4.10 and will be present in the 5.0 release as well.

**Comment by Parakleta** [comment]
I think you should be using one of `substr`, `startsWith`, or `lastIndexOf` as discussed in https://github.com/ReactTraining/history/pull/544#pullrequestreview-78387774 in place of the `indexOf(prefix.toLowerCase())`.

**Pull request closed without merging**